src/raftstore/store/store.rs

```diff
@@ -2661,3 +2634,99 @@ pub fn new_compaction_listener(ch: SendCh<Msg>) -> CompactionListener {
     };
     CompactionListener::new(compacted_handler, Some(size_change_filter))
 }
+
+fn calc_region_declined_bytes(
+    event: CompactedEvent,
+    region_ranges: &BTreeMap<Key, u64>,
+    bytes_threshold: u64,
+) -> Vec<(u64, u64)> {
+    // Calculate influenced regions.
+    let mut influenced_regions = vec![];
+    for (end_key, region_id) in
+        region_ranges.range((Excluded(event.start_key), Included(event.end_key.clone())))
+    {
+        influenced_regions.push((region_id, end_key.clone()));
+    }
+    if let Some((end_key, region_id)) = region_ranges
+        .range((Included(event.end_key), Unbounded))
+        .next()
+    {
+        influenced_regions.push((region_id, end_key.clone()));
+    }
+
+    // Calculate declined bytes for each region.
+    // `end_key` in influenced_regions are in incremental order.
+    let mut region_declined_bytes = vec![];
+    let mut last_end_key: Vec<u8> = vec![];
+    for (region_id, end_key) in influenced_regions.drain(..) {
```

Comment:
It looks like we don't need to `drain`.

Reply:
We need to take away end_key from it.

Reply:
So `for (region_id, end_key) in influenced_regions { .. }`, it is simple and neat.

Reply:
+1